### PR TITLE
feat: ExpressJS support

### PR DESCRIPTION
**Pull request by Mr0Bread**
Express Adapter

### CODE REVIEW

**Comment by Mr0Bread** [comment]
Also would like having this approved first, so that I can make Fastify docs in the manner that we agree on here

**Comment by Mr0Bread** [comment]
@markflorkowski Can you also take a look once you have time?

File: docs/src/pages/_meta.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "getting-started": "Getting Started",
   "nextjs": "Next.js Setup",
   "solid": "SolidStart Setup",
+  "backend-adapters": "Backend Adapters",
```
Comment (juliusmarminge) [inline docs/src/pages/_meta.json:6]: next and solidstart can go under backend adapters too, and then they can link to `/api-reference/react` (and solid)

Comment (juliusmarminge) [reply docs/src/pages/_meta.json:6]: we can do this restructuring later though

File: examples/express/client/package.json
Comment (juliusmarminge) [inline examples/express/client/package.json:1]: unsure about the example

should we ship a vite app to every adapter, or should it just show the backend usage?

File: packages/uploadthing/src/internal/node-http/getBody.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+}): Promise<BodyResult> {
+  const { req, maxBodySize = Infinity } = opts;
+  return new Promise((resolve) => {
+    if ("body" in req) {
+      const isJsonType = req.headers["content-type"] === "application/json";
+
+      if (!isJsonType) {
+        resolve({
+          ok: false,
+          error: new Error("INVALID_CONTENT_TYPE"),
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/node-http/getBody.ts:24]: i think we always wanna throw `UploadThingError`'s

File: examples/express/client/package.json
Comment (Mr0Bread) [reply examples/express/client/package.json:1]: Personally I would ship it with every adapter. In such way it serves not only as example, but also development playground

File: packages/uploadthing/src/internal/node-http/getBody.ts (hunk 2 above)
Comment (Mr0Bread) [reply packages/uploadthing/src/internal/node-http/getBody.ts]: Fixed

File: docs/src/pages/_meta.json (hunk 1 above)
Comment (t3dotgg) [reply docs/src/pages/_meta.json:6]: Don't worry about structure too much until we have the community repo setup, I'll do an overhaul

File: examples/express/client/package.json
Comment (t3dotgg) [reply examples/express/client/package.json:1]: I think we desperately need a separate dev playground. Leave this for now as it's super convenient, I'll prioritize something better for dev in near future

**Review by t3dotgg — APPROVED** [review]

**Review by t3dotgg — COMMENTED** [review]
Wait actually this should probably be a separate package, thoughts @juliusmarminge ?

**Comment by Mr0Bread** [comment]
> Wait actually this should probably be a separate package, thoughts @juliusmarminge ?

If you plan on making adapter for next a separate package, then I guess every adapter, that requires custom code to work, have to be in separate package. For instance, Fastify requires custom adapter and can be moved to separate package

**Comment by juliusmarminge** [comment]
> Wait actually this should probably be a separate package, thoughts @juliusmarminge ?

Why do you think so?

File: packages/uploadthing/src/internal/express.ts [hunk 3]
```diff
@@ -34,16 +32,10 @@ export const createUploadthingExpressHandler = <TRouter extends FileRouter>(
...
-      req: {
-        ...req,
-        // Explicitly passing headers since
-        // without that they are being lost for some reason
-        headers: req.headers,
-        json: () => new Promise((resolve) => resolve(data)),
-      },
+      req: Object.assign(req, {
+        json: () => Promise.resolve(bodyResult.data),
+      }),
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/express.ts:38]: this seems to fix the header situation @Mr0Bread

**Review by juliusmarminge — APPROVED** [review]

File: packages/uploadthing/src/internal/express.ts (hunk 3 above)
Comment (Mr0Bread) [reply packages/uploadthing/src/internal/express.ts:38]: Cool, thank you

**Pull request merged**